File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/core/JavaScriptPropertyScriptValue.kt
package ca.logaritm.dezel.core

/**
 * @class JavaScriptPropertyScriptValue
 * @super JavaScriptPropertyValue
 * @since 0.7.0
 */
open class JavaScriptPropertyScriptValue(value: JavaScriptValue) : JavaScriptPropertyValue() {

	//--------------------------------------------------------------------------
	// Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.7.0
	 */
	init {

		/*
		 * The value was parsed before being assigned here,
		 * there is no need to parse it again.
		 */

		val type: JavaScriptPropertyType

		when (true) {

			value.isNull      -> type = JavaScriptPropertyType.NULL
			value.isUndefined -> type = JavaScriptPropertyType.NULL
			value.isString    -> type = JavaScriptPropertyType.STRING
			value.isNumber    -> type = JavaScriptPropertyType.NUMBER
			value.isBoolean   -> type = JavaScriptPropertyType.BOOLEAN
			value.isObject    -> type = JavaScriptPropertyType.OBJECT
			value.isArray     -> type = JavaScriptPropertyType.ARRAY
			else              -> type = JavaScriptPropertyType.NULL
		}

		this.type = type
		this.unit = JavaScriptPropertyUnit.NONE

		this.value = value
	}

	/**
	 * @method toString
	 * @since 0.7.0
	 */
	override fun toString(): String {
		return this.value!!.string
	}

	/**
	 * @method toNumber
	 * @since 0.7.0
	 */
	override fun toNumber(): Double {
		return this.value!!.number
	}

	/**
	 * @method toBoolean
	 * @since 0.7.0
	 */
	override fun toBoolean(): Boolean {
		return this.value!!.boolean
	}
}

File: platforms/ios/Dezel/Dezel/Core/JavaScriptPropertyScriptValue.swift
/**
 * @class JavaScriptPropertyScriptValue
 * @super JavaScriptPropertyValue
 * @since 0.7.0
 * @hidden
 */
open class JavaScriptPropertyScriptValue : JavaScriptPropertyValue {

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.7.0
	 * @hidden
	 */
	public init(value: JavaScriptValue) {

		/*
		 * The value was parsed before being assigned here,
		 * there is no need to parse it again.
		 */

		let type: JavaScriptPropertyType

		switch (true) {

			case value.isNull:
				type = .null
			case value.isUndefined:
				type = .null
			case value.isString:
				type = .string
			case value.isNumber:
				type = .number
			case value.isBoolean:
				type = .boolean
			case value.isObject:
				type = .object
			case value.isArray:
				type = .array
			case value.isFunction:
				type = .callback

			default:
				type = .null
		}

		super.init(type: type, unit: .none, value: value)
	}

	/**
	 * @method toString
	 * @since 0.7.0
	 */
	override open func toString() -> String {
		return self.value!.string
	}

	/**
	 * @method toNumber
	 * @since 0.7.0
	 */
	override open func toNumber() -> Double {
		return self.value!.number
	}

	/**
	 * @method toBoolean
	 * @since 0.7.0
	 */
	override open func toBoolean() -> Bool {
		return self.value!.boolean
	}
}
